src/HydraAuction/Delegate.hs

```diff
@@ -0,0 +1,107 @@
+module HydraAuction.Delegate (
+  delegateStep,
+  DelegateEvent (..),
+  DelegateInput (..),
+  DelegateRunnerT (..),
+  execDelegateRunnerT,
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Haskell imports
+import Control.Monad.Except (MonadError (throwError))
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.State (MonadState, StateT (..), evalStateT, get, put)
+import Control.Monad.Trans (MonadTrans (lift))
+import Control.Monad.Trans.Except (runExceptT)
+
+-- HydraAuction imports
+
+import HydraAuction.Delegate.Interface (
+  DelegateError (..),
+  DelegateResponse (..),
+  FrontendKind (..),
+  FrontendRequest (..),
+  KnownFrontendRequest (..),
+ )
+import HydraAuction.Hydra.Interface (HydraEvent (..))
+import HydraAuction.Types (AuctionStage (..), AuctionTerms (..))
+
+-- FIXME: add client authentication (where is issue on this)
+
+data DelegateEvent
+  = Start
+  | AuctionStageStarted AuctionStage
+  | HydraEvent HydraEvent
+
+data DelegateInput
+  = DelegateEvent DelegateEvent
+  | FrontendRequest FrontendRequest
+data DelegateState = NoClient | HasClient FrontendKind AuctionTerms
+
+initialState :: DelegateState
+initialState = NoClient
+
+newtype DelegateRunnerT m x = MkDelegateRunner
+  { unDelegateRunner :: StateT DelegateState m x
+  }
+  deriving newtype
+    ( Functor
+    , Applicative
+    , Monad
+    , MonadState DelegateState
+    , MonadIO
+    )
+
+execDelegateRunnerT :: Monad m => DelegateRunnerT m x -> m x
+execDelegateRunnerT (MkDelegateRunner action) = evalStateT action initialState
+
+instance MonadTrans DelegateRunnerT where
+  lift = MkDelegateRunner . lift
+
+delegateStep ::
+  forall m.
+  Monad m =>
+  DelegateInput ->
+  DelegateRunnerT m (Either DelegateError DelegateResponse)
+delegateStep input = runExceptT $ case input of
+  DelegateEvent Start -> return Okay -- FIXME: initialize Hydra
```

Comment:
I would propose to have the delegate server behave similarly to the CLI, in that you need to start the hydra-node before starting the server, and then pass in the http address on which it is listening on.

Reply:
Yes, that is idea. But this is delegated into #158 

The idea is to change `Monad m` into `MonadHydra m` then.

Reply:
Why don't we remove the `Start` event then?

Reply:
Start event is required to Init Hydra head, which does not require to know Auction.

Reply:
Sorry, I thought this was starting the hydra node instead, makes sense now!